### PR TITLE
Add feature : loading unsigned driver

### PR DESCRIPTION
**Pull request by v1k1ngfr**
Hello,
This PR allows to add the unsigned driver loading feature. 
It is done by patching g_CiOptions in order to disable DSE, load unsigned driver, restore g_CiOptions.
The Ci offset is extracted with the same method you use for ntoskrnl.
I added the support of gdrv.sys driver because I'm used to use it.

### CODE REVIEW

**Comment by themaks** [comment]
Thanks a lot for the contribution! Even if not directly linked to EDR evasion, that's a nice feature that could really make testing/research about BYOVD easier!

Out of curiosity : what about KPP/PatchGuard ? I guess patching g_CiOptions is monitored:
- Did it ever trigger any BSOD on your side? 
- Especially did you test leaving it patched for a period longer than the time needed to load an unsigned driver? 
- Also, if the unsigned driver stays loaded after DSE is re-enabled, did you encountered a crash afterwards? (meaning the signature of loaded driver would be periodically checked by CI or KPP)

Don't worry, I'll merge the feature regardless, however, since the tool is sometimes (often?) used in pentests in production environments, we need to warn the user about features that are experimental / dangerous.

Thanks in advance for your answer, I start working on the PR right away.

Cheers

File: EDRSandblast_CLI/EDRSandblast.c [hunk 1]
```diff
@@ -631,6 +657,106 @@ Other options:\n\
...
+                if (g_ciOffsets.st.g_CiOptions == 0x0) {
+                    _putts_or_not(TEXT("[!] The offsets must be computed using the provided script and added to the offsets CSV file (or use --internet). Unsigned driver won't be loaded ...\n"));
+                    lpExitCode = EXIT_FAILURE;
+                }
+                else {
+                    _putts_or_not(TEXT(""));
+                    if (kernelMode) {
+                        DWORD64 CiBaseAddress = 0;
+                        DWORD64 g_CiOptionsAddress = 0;
+                        if (IsCiEnabled() | !IsCiEnabled()) // FIX IT !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
Comment (themaks) [inline EDRSandblast_CLI/EDRSandblast.c:713]: Can this be simply replaced by `IsCiEnabled()` or is there anything else to fix? (like the `IsCiEnabled()` function itself)

Comment (themaks) [reply EDRSandblast_CLI/EDRSandblast.c:713]: It would seem indeed that the function does not work as intended

**Comment by themaks** [comment]
My understanding is that the technique implemented here (patching g_CiOptions) is not working anymore on Windows where KDP is enabled (https://trustedsec.com/blog/g_cioptions-in-a-virtualized-world / https://www.fortinet.com/blog/threat-research/driver-signature-enforcement-tampering), and it's enabled by default at least on Windows 11.

I did not test yet, but I think trying overwriting a KDP-protected value (read only) will result in an instant BSOD, so the current implementation is quite dangerous if it does not check the status of KDP ^^ 

There are some workarounds, as described in the second article, it'll be interesting to check which one is the simplest to implement using the existing basic blocks we already have. PTE remapping is one of them, but I recently read that Microsoft is rolling out new security feature to protect against these attacks

**Comment by themaks** [comment]
Hi @v1k1ngfr ! I merged your work! A quick summary on the changes
* I split the 2 features (GDRV.sys support & DSE bypass) into 2 commits
* Completed the CLI flags to fully use yourDSE bypass feature
* Fixed a systematic crash while trying to restore the old value of g_CiOptions (that probably explains why DSE was permanently disabled after first execution of the tool and you had to add the `if (IsCiEnabled() | !IsCiEnabled())` condition for testing)
* Simplified the way the CI.dll base address is recovered

Also, I think I found the root cause of https://github.com/wavestone-cdt/EDRSandblast/issues/10 (or https://github.com/wavestone-cdt/EDRSandblast/issues/16). The version number mismatch is actually not a problem. In `10.0.19041.2130` / `10.0.19044.2130` for example, the 19041 (resp 19044) number refers to the OS update version (2004, resp. 21H2), but kernel file is identical in both OS, for the same build number (2130). See on https://winbindex.m417z.com/?file=ntoskrnl.exe, you won't find any `10.0.19044.x` version.
I guess the core issue was completely different: if a `ntkrnlmp.pdb` is left in the working directory of EDRSandblast, we use it with on a new kernel version (i.e. Windows updated between two executions) and with the `--internet` option, the old code would try to download the PDB file corresponding to the new kernel version, but abort upon seeing the existing `ntkrnlmp.pdb` file. As a consequence, offsets of a wrong kernel version would be used.

To fix that, I implemented a small PDB parser to identify a GUID (also present in ntoskrnl.exe Debug headers) to validate the PDB is the right one, and re-download it if necessary. See 4d414edb7747da48118a179616a6bbb5bc8dcf81
Hope this definitely closes the bug

Thanks again for your work, and sorry again for the delay ;)
I'll be integrating your work on muting filters soon

Cheers!

**Comment by v1k1ngfr** [comment]
Hey !
I know patching g_CiOptions is monitored on Win10 (I did not encounter crash probably because disabling / loading unsigned driver / re-enabling is done in short time frame, less than a second). 
I also know that Win11 improve this point but I didn't experiment so much on this version.
Story short : yes playing with the kernel memory, g_CiOptions viariables among others , is risky.
Interesting, we should have a talk of other comments, maybe during a security conference ^^

Thanks for merging :) and thanks for sharing ideas & tools

**Pull request closed without merging**